loqrecovery: use local leaseholder status in replica check

aliher1911 opened the pull request:
Previously the rule that is used to pick winning replica looked
on max committed index in the raft log and resoved ties using
store id. This works well in case of 3 replicas and 5+ replicas
if leaseholder is lost (or node with leaseholder restarted).
However for half-online approach not all nodes are restarted and
last leaseholder could be alive and without quorum preventing
recovery.

This commit adds leaseholder status on collected replicas and
then uses previous leaseholders as survivors if all else is
equal. This helps reducing uncertainty and number of nodes that
needs restarting.

Release note: None

Fixes #96264

aliher1911 committed:
loqrecovery: use local leaseholder status in replica check

Previously the rule that is used to pick winning replica looked
on max committed index in the raft log and resoved ties using
store id. This works well in case of 3 replicas and 5+ replicas
if leaseholder is lost (or node with leaseholder restarted).
However for half-online approach not all nodes are restarted and
last leaseholder could be alive and without quorum preventing
recovery.
This commit adds leaseholder status on collected replicas and
then uses previous leaseholders as survivors if all else is
equal. This helps reducing uncertainty and number of nodes that
needs restarting.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/96439)
<!-- Reviewable:end -->

aliher1911 committed:
loqrecovery: report nodes with discarded leaseholders

Previously, if old leaseholder replica was discarded and its
leaseholder prevents recovered replica from becoming leaseholder
and upreplicating, operator has to figure out which nodes needs
to be restarted to unblock the range.
This commit adds an additional list of nodes that needs restarting
to the recovery plan. This list is presented to operator when
plan is staged together with nodes that needs changes applied.

Release note: None

erikgrinaker commented on pkg/cli/debug_recover_loss_of_quorum_test.go:
```diff
@@ -408,14 +408,16 @@ func TestUpdatePlanVsClusterDiff(t *testing.T) {
 	for _, d := range []struct {
 		name    string
 		nodes   []int
+		restart []int
```
nit: consider `staleLease` or some such, because we'll implicitly also be restarting other nodes.

aliher1911 commented on pkg/kv/kvserver/loqrecovery/plan.go:
```diff
@@ -427,6 +439,9 @@ func rankReplicasBySurvivability(replicas []loqrecoverypb.ReplicaInfo) rankedRep
 		if replicas[i].RaftAppliedIndex < replicas[j].RaftAppliedIndex {
 			return false
 		}
+		if replicas[i].LocalAssumesLeaseholder != replicas[j].LocalAssumesLeaseholder {
```
Looks like the note there was only applicable to old loss of quorum recovery where decision had to be made in isolation. Now that we look on all surviving replicas at once, we can pick or not pick a leaseholder and we don't need the determinism mentioned. Which we did. I'll drop the comment altogether.

aliher1911 commented:
bors r=erikgrinaker

aliher1911 commented:
bors r=erikgrinaker
